src/server/worker/pipeline/common.go

```diff
@@ -0,0 +1,135 @@
+package pipeline
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	"github.com/pachyderm/pachyderm/src/server/pkg/backoff"
+	"github.com/pachyderm/pachyderm/src/server/pkg/errutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsconsts"
+	"github.com/pachyderm/pachyderm/src/server/worker/driver"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+)
+
+// RunUserCode will run the pipeline's user code until canceled by the context
+// - used for services and spouts. Unlike how the transform worker runs user
+// code, this does not set environment variables or collect stats.
+func RunUserCode(
+	ctx context.Context,
+	driver driver.Driver,
+	logger logs.TaggedLogger,
+) error {
+	return backoff.RetryUntilCancel(ctx, func() error {
+		// TODO: shouldn't this set up env like the worker does?
+		// TODO: what about the user error handling code?
```

Comment:
@jdoliner

Reply:
Error handling code should be added back in before this is merged.

Reply:
Was that all I'm being @ed over here or is there more clarification needed?